man/man1/txtpar.1
.TH TXTPAR 1NEMO "3 February 2023"

.SH "NAME"
txtpar \- extract parameters from a text file, with optional math

.SH "SYNOPSIS"
.PP
\fBtxtpar in=\fPtext-file \fBp0=\fP \fBp1=\fP ...

.SH "DESCRIPTION"
\fBtxtpar\fP combines some of the often used combinations of
standard Unix text filtering programs like
\fIgrep(1)\fP, \fIawk(1)\fP, and \fIsed(1)\fP, and 
with \fInemoinp(3NEMO)\fP style math into one "simple" NEMO program. Although
\fItabmath(1NEMO)\fP can do some of this processing too, \fItxtpar(1NEMO)\fP's strength is
processing multiple lines in a reduction style, instead of filtering. It often simplifies
the esoteric syntax used in the grep/awk/sed pipes, but is much more limited.
.PP
.nf
     grep RA rsr.33551.driver.sum.txt  | head -1 | awk '{print $4}'
vs.
     txtpar rsr.33551.driver.sum.txt p0=RA,1,4
.fi
.PP
Normally one parameter is extracted per \fBp#=\fP directive, in a \fIword,row,col\fP fashion. Future option
will allow multiple \fIcol's\fP , but this is not implemented yet.
.PP
Comment lines in the input file are considered part of the text, they are not stripped out. Use
\fItabcomment(1NEMO)\fP first if you really need them removed.
.PP
Optionally additional math is allowed on the extracted parameters using \fInemoinp(1NEMO)\fP style expressions. In
this case, the parameters themselves are not output, but only the ones computed via \fBexpr=\fP. See EXAMPLES below
if you need both.
.PP
To extract values from a \fIkey=val\fP style parameter file, use \fInemopars(8NEMO)\fP.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fIin-file\fP
input text file, containing marker words and numeric values to be parsed. 
No default.
.TP
\fBexpr=\fP\fIexpr1,...\fP
Optional expression(s), separated by commas, referencing the extracted parameters via the \fIfie(3NEMO)\fP parsing rules.
By default no expression is needed and each parameter is output as they appeared in the text file, however when \fBexpr=\fP is used, only those
newly computed values are output.  [Default: no expression given]
.TP
\fBformat=\fP\fIexpression\fP
A (\fIprintf(3)\fP) format specification applied to the computed expressions
with which the parameters are written to output [Default: \fB%g\fP].
.TP
\fBseed=\fP\fIseed\fP
Integer initial seed in case random numbers have been used in the expression.
If 0 is given, the time of the day will be used (see 
\fIxrandom(3NEMO)\fP for other special seed values)
to initialize the random number generator. [Default: \fB0\fP].
.TP
\fBnewline=t|f\fP
Should a newline be used between output values. [Default: \fBf\fP]
.TP
\fBmaxline=\fP\fImax\fP
In case a NEMO/unix pipe is used, the maximum number of lines will need to be specified. [Default: 10000]
.TP
\fBp#=\fP\fI[word,]row,col,...\fP
Parameter selection directive in the indexed keyword notation (e.g. p0=, p1=, p2=), see \fIgetparam(3NEMO)\fP.
\fIword\fP is the string to search on (with some limited grep syntax), but can be blank if all lines need to apply.
\fIrow\fP is an integer, and is the row-th occurance from the \fIword\fP match. 1 would be the first, 2 the 2nd, -1 the last, -2 the one but last etc.
0 is not an allowed row number.
\fIcol\fP is an integer, and designates the "column" to look for in the current row (line), 1 would be the first word, -1 the last etc.
Multiple \fIcol\fP will be allowed in a future version, but currently not available. 0 is not an allowed col number.

At least one parameter needs to be given: \fBp0=\fP.

.SH "EXAMPLES"
Here is an example of extracting three values and converting them to two output values:
.nf

       txtpar in=example.txt expr="log(abs(%1)),log(abs(%3/%2))" format=%.3f \\
              p0=Worst,1,9  p1=QAC,1,3 p2=QAC,1,4
	
       cat example.txt
       \fIWorst fractional energy loss dE/E = (E_t-E_0)/E_0 = 0.00146761 at T = 1.28125
       QAC_STATS: - 0.039966 0.0274195 0.00185505 0.135854  0.799319 1  20
       \fP

.fi

.PP
If both original values as well as newly computed \fBexpr=\fP are desired in the output (this is the default
behavior in \fItabmath(1NEMO)\fP), an extra pipe using
\fItabmath(1NEMO)\fP) is needed. Here is an example to add the first column in the first two lines of a table
in both forms, with and without the original values
.nf

      txtpar example.txt p0=1,1 p1=2,1 | tabmath - - %1+%2 format=%.3f
      txtpar example.txt %1+%2 %.3f p0=1,1 p1=2,1 

.fi

.SH "CAVEATS"
This program is not intended for streaming very large files, as the file is read into memory for processing
and each parameter could potentially cause a search through the file.

.SH "TODO"
On the wishlist are relative row/col w.r.t. the matched word using an explicit + or - in front
of the row or col:
.nf
     p0=word,+row,col
     p0=word,row,-col
.fi     

.SH "SEE ALSO"
grep(1), awk(1), nemopars(8NEMO), tabmath(1NEMO), tabrows(1NEMO), tabcomment(1NEMO), table(5NEMO), run(1NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.0i +4.0i
27-dec-2021	V0.1 simple draft written - implemented simple version	PJT
.fi


src/kernel/tab/txtpar.c
/*
 * TXTPAR: grab parameters from a text file
 *
 *     27-dec-2021    drafted
 *
 *  @todo   bring back comments
 */

/**************** INCLUDE FILES ********************************/ 


#include <stdinc.h>
#include <getparam.h>
#include <table.h>
#include <extstring.h>
#include <ctype.h>

/**************** COMMAND LINE PARAMETERS **********************/

string defv[] = {                /* DEFAULT INPUT PARAMETERS */
    "in=???\n           input file name(s)",
    "expr=\n            formula for parameters after extraction",
    "format=%g\n        format for new output values",
    "seed=0\n           Initial random number",
    "newline=f\n        add newline between output parameters",
#if 0    
    "maxline=10000\n    Max number of lines in case a pipe was used",
#endif
    "p#=\n              The word,row,col tuples for given parameter",
    "VERSION=0.7b\n     18-may-2022 PJT",
    NULL
};

string usage = "Extract numeric parameters from a text file, with optional math";



/**************** GLOBAL VARIABLES *******************************/

string input;			        /* file names */
stream instr;			        /* file streams */
table *tptr;                            /* table */

string fmt;                             /* format of new column */

#define MAXCOL          256             /* MAXIMUM number of columns */
#define MAXROW          256
#define MLINELEN       8196		/* linelength of catenated */
#define MAXPAR          100

bool   keepc[MAXCOL+1];                 /* columns to keep (t/f) */
int    ndelc;                           /* actual number of skip columns */

string *fies, selfie;                   /* fie pointers */
int    nfies;                           /* number of fie pointers */
bool   Qfie;				/* boolean if multiple fie's loaded */
bool   Qnewline;                        /* boolean if newline is needed */
bool   Qexpr;

int    nlines;

string word[MAXPAR];
int    row[MAXPAR];
int    col[MAXPAR];

int    maxpar;

local void setparams(void);
local void convert(stream);
local string *burstfie(string);

extern  string *burststring(string, string);
extern  int inifie(string);
extern void dofie(real *, int *, real *, real *);
extern void dmpfie(void);
extern int savefie(int);
extern int loadfie(int);


void nemo_main(void)
{
    setparams();
    convert(instr);
}

local void setparams(void)
{
    string newcol;                          /* formula for new column */
    string delcol;                          /* which columns not to write */
    int    delc[MAXCOL];                    /* columns to skip for output */
    int    wrc[3];                          /* hold "word","row","col" - temp */
    int i, nwrc, ispar, nsp;
    string *sp;
    char   *cp;

    input  = getparam("in");
    instr = stropen(input,"r");
    tptr = table_open(instr,-1);            /* open & read table, treating comments like data */

    Qexpr = hasvalue("expr");
    newcol = getparam("expr");
    fmt = getparam("format");

    // check all indexed parameters
    maxpar = indexparam("p",-1);
    if (maxpar < 0) error("No parameter directives (p0=, p1, ...) given");
    for (i=0; i<maxpar; i++) {
      ispar = indexparam("p",i);
      if (ispar) {
	dprintf(2,"%d : %s\n", i, getparam_idx("p",i));
	cp = getparam_idx("p",i);
	if (*cp == ',') {
	  sp = burststring(cp+1, ",");
	  nsp = xstrlen(sp, sizeof(string)) - 1;
	  if (nsp < 2 ) error("Need row,col");
	  word[i] = strdup("");
	  row[i] = atoi(sp[0]);
	  col[i] = atoi(sp[1]);
	} else {
	  sp = burststring(cp, ",");	  
	  nsp = xstrlen(sp, sizeof(string)) - 1;
	  if (nsp == 2) {    // this could be a bad idea if we allow multiple col's
	    word[i] = strdup("");
	    row[i] = atoi(sp[0]);
	    col[i] = atoi(sp[1]);
	  } else if (nsp == 3) {
	    word[i] = strdup(sp[0]);
	    row[i] = atoi(sp[1]);
	    col[i] = atoi(sp[2]);
	  } else
	      error("Need word,row,col");
	}
      } else {
	if (i==0) warning("p0 is missing");
	word[i] = NULL;
      }
    }

    fies = burstfie(newcol);
    nfies = xstrlen(fies,sizeof(string)) - 1;
    if(nfies)dprintf(1,"%d functions to parse\n",nfies);
    for (i=0; i<nfies; i++) {
	dprintf(1,"Saving: %s\n",fies[i]);
        inifie(fies[i]);
        if (savefie(i+1) < 0) error("Could not save fie[%d]: %s\n",i,fies[i]);
	if(nemo_debug(1)) dmpfie();
    }
    Qfie = nfies > 1;

    init_xrandom(getparam("seed"));
    Qnewline = getbparam("newline");
}



local void convert(stream instr)
{
    char   line[MLINELEN];
    real   dval[MAXCOL];            /* number of items (values on line) */
    string sval[MAXCOL];            // 
    int    nval, i, j, one=1;
    int    match[MAXROW], nmatch, rownr;
    string *outv;                   /* pointer to vector of strings to write */
    char   *cp, *seps=", \t";       /* column separators  */
    real   errval=0.0;

    /*
     *   read input lines
     */

    nlines = table_nrows(tptr);

    /*
     *   extract parameters
     */

    nval = 0;       // this will fill the dval[]
    dprintf(1,"Parsing %d parameters p#=\n", maxpar);
    for (i=0; i<maxpar; i++) {                          // loop all parameters
      if (strlen(word[i]) > 0) {                        // check if a word match was needed
	dprintf(1,"Searching word=%s\n",word[i]);
	nmatch = 0;
	for (j=0; j<nlines; j++) {
	  // @todo  allow ^word 
	  if (strstr(table_row(tptr,j),word[i])) {
	    dprintf(1,"Match in %d: %s\n",j,table_row(tptr,j));
	    match[nmatch++] = j;
	  }
	}
	if (nmatch == 0) {
	  error("No match found for \"%s\" in %s", word[i],input);
	  continue;
	}
      } else
	nmatch = 0;
      dprintf(1,"nmatch: %d\n",nmatch);
      
      if (nmatch == 0) {                        // can use absolute row number
	if (row[i] > 0)
	  rownr = row[i]-1;
	else if (row[i] < 0) 
	  rownr = nlines+row[i];
	else
	  error("Illegal row=0 reference - 1");
	if (rownr < 0 || rownr >= nlines)
	  error("bad rownumber %d for p%d in %s", row[i],i,input);
      } else {                                  // look in matched values
	if (row[i] > 0) {
	  if (row[i] > nmatch) error("Not enough matches for %s,%d in %s",word[i],row[i],input);
	  rownr = match[row[i]-1];
	} else if (row[i] < 0) {
	  if (nmatch+row[i] < 0) error("Not enough matches for %s,%d in %s",word[i],row[i],input);	  
	  rownr = match[nmatch+row[i]];
	} else
	  error("Illegal row=0 reference - 2 in %s",input);
      }
      
      dprintf(1,"par p%d row=%d col=%d\n", i, rownr, col[i]);
      cp = table_row(tptr,rownr);
      dprintf(1,"LINE: %s\n", cp);
      outv = burststring(cp, seps);
      sval[nval] = outv[col[i]-1];
      nval++;
    }
    if (nval == 0) warning("no output??? should never happen");


    /*
     *  if no expressions give, output parameters "as is" from their textual input
     */

    if (!Qexpr) {
      // simple output of input parameters
      for (i=0; i<nval; i++) {
	if (i>0 && !Qnewline) printf(" ");
	printf("%s", sval[i]);
	if (Qnewline) printf("\n");
      }
      if (!Qnewline) printf("\n");
      return;
    }

    /*
     *   convert strings to float values
     */
    
    for (i=0; i<nval; i++)
      dval[i] = atof(sval[i]);

    /*
     *   apply math and output the new expressions
     */

    line[0] = '\0';
    for(i=0; i<nfies; i++) {
      if (Qfie) loadfie(i+1);
      dofie(dval,&one,&dval[nval+i],&errval);
      dprintf(3," dofie(%d) -> %g  %g\n",i+1,dval[nval+i], errval); //BUG
      strcat(line," ");
      printf(fmt,dval[nval+i]);
      if (nfies>1) printf(" ");
      if (Qnewline)
	printf("\n");
    }
    if (!Qnewline)
      printf("\n");
    
}

/* burstfie(): to be placed with burststring() later on...
 *
 *	18-feb-92	written		PJT
 */

#define MWRD  256	/* max words in list */
#define MSTR  256	/* max chars per word */

local string *burstfie(string lst)
{
    string wrdbuf[MWRD], *wp;
    char strbuf[MSTR], *sp, *lp;
    int level=0;

    wp = wrdbuf;
    sp = strbuf;
    lp = lst;
    do {						/* scan over list */
	if (*lp == 0 || (*lp == ',' && level == 0)){	/*   is this a sep? */
	    if (sp > strbuf) {				/*     got a word? */
		*sp = 0;
		*wp++ = (string) copxstr(strbuf, sizeof(char));
		if (wp == &wrdbuf[MWRD])		/*       no room? */
		    error("burststring: too many words");
		sp = strbuf;				/*       for next 1 */
	    }
	} else {					/*   part of word */
            if (*lp == '(') level++;
            if (*lp == ')') level--;
	    *sp++ = *lp;				/*     so copy it */
	    if (sp == &strbuf[MSTR])			/*     no room? */
		error("burststring: word too long");
	}
    } while (*lp++ != 0);				/* until list ends */
    *wp = 0;
    return ((string *) copxstr(wrdbuf, sizeof(string)));	/*PPAP*/
}

